// @akord/akord-js/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.functions = exports.smartweaveValues = exports.smartweaveTags = exports.encryptionTags = exports.fileTags = exports.dataTags = exports.protocolTags = exports.AKORD_TAG = exports.actionRefs = exports.role = exports.status = exports.objectType = exports.reactionEmoji = void 0;
var reactionEmoji;
(function (reactionEmoji) {
    reactionEmoji["JOY"] = "128514";
    reactionEmoji["ASTONISHED"] = "128562";
    reactionEmoji["CRY"] = "128557";
    reactionEmoji["HEART"] = "10084,65039";
    reactionEmoji["FIRE"] = "128293";
    reactionEmoji["THUMBS_UP"] = "128077";
    reactionEmoji["THUMBS_DOWN"] = "128078";
    reactionEmoji["PRAY"] = "128591";
})(reactionEmoji = exports.reactionEmoji || (exports.reactionEmoji = {}));
;
var objectType;
(function (objectType) {
    objectType["VAULT"] = "Vault";
    objectType["MEMBERSHIP"] = "Membership";
    objectType["STACK"] = "Stack";
    objectType["MEMO"] = "Memo";
    objectType["FOLDER"] = "Folder";
    objectType["NOTE"] = "Note";
    objectType["PROFILE"] = "Profile";
    objectType["NFT"] = "NFT";
})(objectType = exports.objectType || (exports.objectType = {}));
;
var status;
(function (status) {
    status["PENDING"] = "PENDING";
    status["INVITED"] = "INVITED";
    status["ACCEPTED"] = "ACCEPTED";
    status["REJECTED"] = "REJECTED";
    status["REVOKED"] = "REVOKED";
    status["ACTIVE"] = "ACTIVE";
    status["ARCHIVED"] = "ARCHIVED";
    status["DELETED"] = "DELETED";
})(status = exports.status || (exports.status = {}));
;
var role;
(function (role) {
    role["OWNER"] = "OWNER";
    role["CONTRIBUTOR"] = "CONTRIBUTOR";
    role["VIEWER"] = "VIEWER";
})(role = exports.role || (exports.role = {}));
;
var actionRefs;
(function (actionRefs) {
    actionRefs["VAULT_CREATE"] = "VAULT_CREATE";
    actionRefs["VAULT_RENAME"] = "VAULT_RENAME";
    actionRefs["VAULT_UPDATE_METADATA"] = "VAULT_UPDATE_METADATA";
    actionRefs["VAULT_ADD_TAGS"] = "VAULT_ADD_TAGS";
    actionRefs["VAULT_REMOVE_TAGS"] = "VAULT_REMOVE_TAGS";
    actionRefs["VAULT_ARCHIVE"] = "VAULT_ARCHIVE";
    actionRefs["VAULT_RESTORE"] = "VAULT_RESTORE";
    actionRefs["VAULT_DELETE"] = "VAULT_DELETE";
    actionRefs["MEMBERSHIP_INVITE"] = "MEMBERSHIP_INVITE";
    actionRefs["MEMBERSHIP_INVITE_RESEND"] = "MEMBERSHIP_INVITE_RESEND";
    actionRefs["MEMBERSHIP_ACCEPT"] = "MEMBERSHIP_ACCEPT";
    actionRefs["MEMBERSHIP_REVOKE"] = "MEMBERSHIP_REVOKE";
    actionRefs["MEMBERSHIP_REJECT"] = "MEMBERSHIP_REJECT";
    actionRefs["MEMBERSHIP_LEAVE"] = "MEMBERSHIP_LEAVE";
    actionRefs["MEMBERSHIP_KEY_ROTATE"] = "MEMBERSHIP_KEY_ROTATE";
    actionRefs["MEMBERSHIP_CHANGE_ROLE"] = "MEMBERSHIP_CHANGE_ACCESS";
    actionRefs["MEMBERSHIP_PROFILE_UPDATE"] = "MEMBERSHIP_PROFILE_UPDATE";
    actionRefs["MEMBERSHIP_OWNER"] = "MEMBERSHIP_OWNER";
    actionRefs["MEMBERSHIP_CONFIRM"] = "MEMBERSHIP_CONFIRM";
    actionRefs["MEMO_CREATE"] = "MEMO_CREATE";
    actionRefs["MEMO_ADD_REACTION"] = "MEMO_ADD_REACTION";
    actionRefs["MEMO_REMOVE_REACTION"] = "MEMO_REMOVE_REACTION";
    actionRefs["FOLDER_CREATE"] = "FOLDER_CREATE";
    actionRefs["FOLDER_RENAME"] = "FOLDER_RENAME";
    actionRefs["FOLDER_MOVE"] = "FOLDER_MOVE";
    actionRefs["FOLDER_REVOKE"] = "FOLDER_REVOKE";
    actionRefs["FOLDER_RESTORE"] = "FOLDER_RESTORE";
    actionRefs["FOLDER_DELETE"] = "FOLDER_DELETE";
    actionRefs["STACK_CREATE"] = "STACK_CREATE";
    actionRefs["STACK_UPLOAD_REVISION"] = "STACK_UPLOAD_REVISION";
    actionRefs["STACK_RENAME"] = "STACK_RENAME";
    actionRefs["STACK_MOVE"] = "STACK_MOVE";
    actionRefs["STACK_REVOKE"] = "STACK_REVOKE";
    actionRefs["STACK_RESTORE"] = "STACK_RESTORE";
    actionRefs["STACK_DELETE"] = "STACK_DELETE";
    actionRefs["NOTE_CREATE"] = "NOTE_CREATE";
    actionRefs["NOTE_UPLOAD_REVISION"] = "NOTE_UPLOAD_REVISION";
    actionRefs["NOTE_RENAME"] = "NOTE_RENAME";
    actionRefs["NOTE_MOVE"] = "NOTE_MOVE";
    actionRefs["NOTE_REVOKE"] = "NOTE_REVOKE";
    actionRefs["NOTE_RESTORE"] = "NOTE_RESTORE";
    actionRefs["NOTE_DELETE"] = "NOTE_DELETE";
    actionRefs["PROFILE_UPDATE"] = "PROFILE_UPDATE";
    actionRefs["NFT_MINT"] = "NFT_MINT";
})(actionRefs = exports.actionRefs || (exports.actionRefs = {}));
;
exports.AKORD_TAG = "Akord-Tag";
var protocolTags;
(function (protocolTags) {
    protocolTags["CLIENT_NAME"] = "Client-Name";
    protocolTags["PROTOCOL_NAME"] = "Protocol-Name";
    protocolTags["PROTOCOL_VERSION"] = "Protocol-Version";
    protocolTags["TIMESTAMP"] = "Timestamp";
    protocolTags["FUNCTION_NAME"] = "Function-Name";
    protocolTags["VAULT_ID"] = "Vault-Id";
    protocolTags["MEMBERSHIP_ID"] = "Membership-Id";
    protocolTags["NODE_TYPE"] = "Node-Type";
    protocolTags["NODE_ID"] = "Node-Id";
    protocolTags["PUBLIC"] = "Public";
    protocolTags["REF_ID"] = "Ref-Id";
    protocolTags["REVISION"] = "Revision";
    protocolTags["ACTION_REF"] = "Action-Ref";
    protocolTags["GROUP_REF"] = "Group-Ref";
    protocolTags["SIGNER_ADDRESS"] = "Signer-Address";
    protocolTags["SIGNATURE"] = "Signature";
    protocolTags["MEMBER_ADDRESS"] = "Member-Address";
    protocolTags["PARENT_ID"] = "Parent-Id";
})(protocolTags = exports.protocolTags || (exports.protocolTags = {}));
;
var dataTags;
(function (dataTags) {
    dataTags["DATA_TYPE"] = "Data-Type";
    dataTags["CONTENT_TYPE"] = "Content-Type";
    dataTags["SIGNER_ADDRESS"] = "Signer-Address";
    dataTags["SIGNATURE"] = "Signature";
})(dataTags = exports.dataTags || (exports.dataTags = {}));
;
var fileTags;
(function (fileTags) {
    fileTags["FILE_NAME"] = "File-Name";
    fileTags["FILE_MODIFIED_AT"] = "File-Modified-At";
    fileTags["FILE_SIZE"] = "File-Size";
    fileTags["FILE_TYPE"] = "File-Type";
    fileTags["FILE_HASH"] = "File-Hash";
})(fileTags = exports.fileTags || (exports.fileTags = {}));
var encryptionTags;
(function (encryptionTags) {
    encryptionTags["IV"] = "Initialization-Vector";
    encryptionTags["ENCRYPTED_KEY"] = "Encrypted-Key";
    encryptionTags["PUBLIC_ADDRESS"] = "Public-Address";
})(encryptionTags = exports.encryptionTags || (exports.encryptionTags = {}));
;
var smartweaveTags;
(function (smartweaveTags) {
    smartweaveTags["APP_NAME"] = "App-Name";
    smartweaveTags["APP_VERSION"] = "App-Version";
    smartweaveTags["CONTENT_TYPE"] = "Content-Type";
    smartweaveTags["CONTRACT"] = "Contract";
    smartweaveTags["CONTRACT_SOURCE"] = "Contract-Src";
    smartweaveTags["INIT_STATE"] = "Init-State";
    smartweaveTags["INPUT"] = "Input";
    smartweaveTags["INTERACT_WRITE"] = "Interact-Write";
})(smartweaveTags = exports.smartweaveTags || (exports.smartweaveTags = {}));
;
var smartweaveValues;
(function (smartweaveValues) {
    smartweaveValues["CONTRACT_CODE_SOURCE"] = "SmartWeaveContractSource";
    smartweaveValues["CONTRACT_INTERACTION"] = "SmartWeaveAction";
    smartweaveValues["CONTRACT_INITIALIZATION"] = "SmartWeaveContract";
})(smartweaveValues = exports.smartweaveValues || (exports.smartweaveValues = {}));
;
var functions;
(function (functions) {
    functions["VAULT_CREATE"] = "vault:init";
    functions["VAULT_UPDATE"] = "vault:update";
    functions["VAULT_ARCHIVE"] = "vault:archive";
    functions["VAULT_RESTORE"] = "vault:restore";
    functions["MEMBERSHIP_INVITE"] = "membership:invite";
    functions["MEMBERSHIP_ACCEPT"] = "membership:accept";
    functions["MEMBERSHIP_REVOKE"] = "membership:revoke";
    functions["MEMBERSHIP_REJECT"] = "membership:reject";
    functions["MEMBERSHIP_CHANGE_ROLE"] = "membership:change-role";
    functions["MEMBERSHIP_UPDATE"] = "membership:update";
    functions["MEMBERSHIP_ADD"] = "membership:add";
    functions["NODE_CREATE"] = "node:create";
    functions["NODE_UPDATE"] = "node:update";
    functions["NODE_REVOKE"] = "node:revoke";
    functions["NODE_MOVE"] = "node:move";
    functions["NODE_RESTORE"] = "node:restore";
    functions["NODE_DELETE"] = "node:delete";
})(functions = exports.functions || (exports.functions = {}));
;
