// @akord/akord-js/lib/api/akord-api.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AkordApi = exports.defaultFileUploadOptions = void 0;
const api_1 = require("./api");
const config_1 = require("./config");
const api_client_1 = require("./api-client");
const logger_1 = require("../logger");
exports.defaultFileUploadOptions = {
    cacheOnly: false,
    public: false
};
class AkordApi extends api_1.Api {
    constructor(config) {
        super();
        this.config = (0, config_1.apiConfig)(config.env);
    }
    uploadData(items, options = exports.defaultFileUploadOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const resources = [];
            yield Promise.all(items.map((item, index) => __awaiter(this, void 0, void 0, function* () {
                const resource = yield new api_client_1.ApiClient()
                    .env(this.config)
                    .data({ data: item.data, tags: item.tags })
                    .cacheOnly(options.cacheOnly)
                    .uploadState();
                logger_1.Logger.log("Uploaded state with id: " + resource);
                resources[index] = resource;
            })));
            return resources;
        });
    }
    ;
    postContractTransaction(vaultId, input, tags, metadata) {
        return __awaiter(this, void 0, void 0, function* () {
            const { id, object } = yield new api_client_1.ApiClient()
                .env(this.config)
                .vaultId(vaultId)
                .metadata(metadata)
                .input(input)
                .tags(tags)
                .transaction();
            logger_1.Logger.log("Uploaded contract interaction with id: " + id);
            return { id, object };
        });
    }
    ;
    getMembers(vaultId) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield new api_client_1.ApiClient()
                .env(this.config)
                .vaultId(vaultId)
                .getMembers();
        });
    }
    ;
    initContractId(tags, state) {
        return __awaiter(this, void 0, void 0, function* () {
            const contractId = yield new api_client_1.ApiClient()
                .env(this.config)
                .data({ tags, state })
                .contract();
            logger_1.Logger.log("Created contract with id: " + contractId);
            return contractId;
        });
    }
    ;
    uploadFile(file, tags, options = exports.defaultFileUploadOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const uploadOptions = Object.assign(Object.assign({}, exports.defaultFileUploadOptions), options);
            const resource = yield new api_client_1.ApiClient()
                .env(this.config)
                .data(file)
                .tags(tags)
                .public(uploadOptions.public)
                .cacheOnly(uploadOptions.cacheOnly)
                .progressHook(uploadOptions.progressHook)
                .cancelHook(uploadOptions.cancelHook)
                .uploadFile();
            logger_1.Logger.log("Uploaded file with id: " + resource.id);
            return resource;
        });
    }
    ;
    downloadFile(id, options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const { response } = yield new api_client_1.ApiClient()
                .env(this.config)
                .resourceId(id)
                .public(options.public)
                .numberOfChunks(options.numberOfChunks)
                .progressHook(options.progressHook, options.loadedSize, options.resourceSize)
                .cancelHook(options.cancelHook)
                .asArrayBuffer()
                .downloadFile();
            const fileData = response.data;
            const metadata = { encryptedKey: response.headers["x-amz-meta-encryptedkey"], iv: response.headers["x-amz-meta-iv"] };
            return { fileData, metadata };
        });
    }
    ;
    existsUser(email) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield new api_client_1.ApiClient()
                .env(this.config)
                .queryParams({ email })
                .existsUser();
        });
    }
    getUserPublicData(email) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield new api_client_1.ApiClient()
                .env(this.config)
                .queryParams({ email })
                .getUserPublicData();
        });
    }
    ;
    getUser() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield new api_client_1.ApiClient()
                .env(this.config)
                .getUser();
        });
    }
    ;
    updateUser(name, avatarUri) {
        return __awaiter(this, void 0, void 0, function* () {
            yield new api_client_1.ApiClient()
                .env(this.config)
                .data({
                name: name,
                avatarUri: avatarUri
            })
                .updateUser();
        });
    }
    ;
    deleteVault(vaultId) {
        return __awaiter(this, void 0, void 0, function* () {
            yield new api_client_1.ApiClient()
                .env(this.config)
                .vaultId(vaultId)
                .deleteVault();
        });
    }
    inviteNewUser(vaultId, email, role, message) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield new api_client_1.ApiClient()
                .env(this.config)
                .vaultId(vaultId)
                .data({
                email: email,
                role: role
            })
                .metadata({ message })
                .invite();
        });
    }
    revokeInvite(vaultId, membershipId) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield new api_client_1.ApiClient()
                .env(this.config)
                .vaultId(vaultId)
                .resourceId(membershipId)
                .revokeInvite();
        });
    }
    inviteResend(vaultId, membershipId) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield new api_client_1.ApiClient()
                .env(this.config)
                .vaultId(vaultId)
                .resourceId(membershipId)
                .inviteResend();
        });
    }
    getNode(id, type) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield new api_client_1.ApiClient()
                .env(this.config)
                .resourceId(id)
                .queryParams({ type })
                .getNode();
        });
    }
    ;
    getMembership(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield new api_client_1.ApiClient()
                .env(this.config)
                .resourceId(id)
                .getMembership();
        });
    }
    ;
    getVault(id, options) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield new api_client_1.ApiClient()
                .env(this.config)
                .resourceId(id)
                .queryParams({
                withNodes: options === null || options === void 0 ? void 0 : options.withNodes,
                withMemberships: options === null || options === void 0 ? void 0 : options.deep,
                withMemos: options === null || options === void 0 ? void 0 : options.deep,
                withStacks: options === null || options === void 0 ? void 0 : options.deep,
                withFolders: options === null || options === void 0 ? void 0 : options.deep,
            })
                .getVault();
        });
    }
    ;
    getMembershipKeys(vaultId) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield new api_client_1.ApiClient()
                .env(this.config)
                .vaultId(vaultId)
                .getMembershipKeys();
        });
    }
    ;
    getNodeState(stateId) {
        return __awaiter(this, void 0, void 0, function* () {
            const { response } = yield new api_client_1.ApiClient()
                .env(this.config)
                .resourceId(stateId)
                .downloadState();
            return response.data;
        });
    }
    ;
    getNotifications() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield new api_client_1.ApiClient()
                .env(this.config)
                .getNotifications();
        });
    }
    ;
    readNotifications(options) {
        return __awaiter(this, void 0, void 0, function* () {
            yield new api_client_1.ApiClient()
                .env(this.config)
                .queryParams(options)
                .patchNotifications();
        });
    }
    ;
    getContractState(objectId) {
        return __awaiter(this, void 0, void 0, function* () {
            const contract = yield new api_client_1.ApiClient()
                .env(this.config)
                .vaultId(objectId)
                .getContract();
            return contract.state;
        });
    }
    ;
    getMemberships(options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield new api_client_1.ApiClient()
                .env(this.config)
                .queryParams({
                limit: options.limit,
                nextToken: options.nextToken
            })
                .getMemberships();
        });
    }
    ;
    getVaults(options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield new api_client_1.ApiClient()
                .env(this.config)
                .queryParams({
                tags: JSON.stringify(options.tags ? options.tags : {}),
                filter: JSON.stringify(options.filter ? options.filter : {}),
                limit: options.limit,
                nextToken: options.nextToken
            })
                .getVaults();
        });
    }
    ;
    getNodesByVaultId(vaultId, type, options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield new api_client_1.ApiClient()
                .env(this.config)
                .vaultId(vaultId)
                .queryParams({
                type,
                parentId: options.parentId,
                tags: JSON.stringify(options.tags ? options.tags : {}),
                filter: JSON.stringify(options.filter ? options.filter : {}),
                limit: options.limit,
                nextToken: options.nextToken
            })
                .getNodesByVaultId();
        });
    }
    ;
    getMembershipsByVaultId(vaultId, options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield new api_client_1.ApiClient()
                .env(this.config)
                .vaultId(vaultId)
                .queryParams({
                filter: JSON.stringify(options.filter ? options.filter : {}),
                limit: options.limit,
                nextToken: options.nextToken
            })
                .getMembershipsByVaultId();
        });
    }
    ;
    getTransactions(vaultId) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield new api_client_1.ApiClient()
                .env(this.config)
                .vaultId(vaultId)
                .getTransactions();
        });
    }
}
exports.default = AkordApi;
exports.AkordApi = AkordApi;
