// @akord/akord-js/lib/types/membership.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Membership = void 0;
const crypto_1 = require("@akord/crypto");
const profile_details_1 = require("./profile-details");
class Membership extends crypto_1.Encryptable {
    constructor(membershipProto, keys) {
        super(keys, null);
        this.id = membershipProto.id;
        this.owner = membershipProto.owner;
        this.address = membershipProto.address;
        this.createdAt = membershipProto.createdAt;
        this.updatedAt = membershipProto.updatedAt;
        this.expiresAt = membershipProto.expiresAt;
        this.data = membershipProto.data;
        this.status = membershipProto.status;
        this.role = membershipProto.role;
        this.encPublicSigningKey = membershipProto.encPublicSigningKey;
        this.email = membershipProto.email;
        this.memberPublicSigningKey = membershipProto.memberPublicSigningKey;
        this.vaultId = membershipProto.vaultId;
        this.keys = membershipProto.keys;
        this.memberDetails = new profile_details_1.ProfileDetails(membershipProto.memberDetails, keys);
        this.__public__ = membershipProto.__public__;
        this.__cacheOnly__ = membershipProto.__cacheOnly__;
    }
}
exports.Membership = Membership;
