// @akord/akord-js/lib/types/vault.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Vault = void 0;
const crypto_1 = require("@akord/crypto");
const membership_1 = require("./membership");
const node_1 = require("./node");
class Vault extends crypto_1.Encryptable {
    constructor(vaultProto, keys) {
        var _a, _b, _c, _d;
        super(keys, null);
        this.id = vaultProto.id;
        this.owner = vaultProto.owner;
        this.public = vaultProto.public;
        this.createdAt = vaultProto.createdAt;
        this.updatedAt = vaultProto.updatedAt;
        this.size = vaultProto.size;
        this.name = vaultProto.name;
        this.description = vaultProto.description;
        this.termsOfAccess = vaultProto.termsOfAccess;
        this.tags = vaultProto.tags;
        this.status = vaultProto.status;
        this.data = vaultProto.data;
        this.cacheOnly = vaultProto.cacheOnly;
        this.tags = vaultProto.tags;
        this.memberships = (_a = vaultProto === null || vaultProto === void 0 ? void 0 : vaultProto.memberships) === null || _a === void 0 ? void 0 : _a.map((membership) => new membership_1.Membership(membership, keys));
        this.memos = (_b = vaultProto === null || vaultProto === void 0 ? void 0 : vaultProto.memos) === null || _b === void 0 ? void 0 : _b.map((memo) => new node_1.Memo(memo, keys));
        this.stacks = (_c = vaultProto === null || vaultProto === void 0 ? void 0 : vaultProto.stacks) === null || _c === void 0 ? void 0 : _c.map((stack) => new node_1.Stack(stack, keys));
        this.folders = (_d = vaultProto === null || vaultProto === void 0 ? void 0 : vaultProto.folders) === null || _d === void 0 ? void 0 : _d.map((folder) => new node_1.Folder(folder, keys));
    }
}
__decorate([
    (0, crypto_1.encrypted)()
], Vault.prototype, "name", void 0);
__decorate([
    (0, crypto_1.encrypted)()
], Vault.prototype, "description", void 0);
exports.Vault = Vault;
